plutus-core/plutus-ir/src/PlutusIR/Transform/ThunkRecursions.hs

```diff
@@ -93,51 +94,74 @@ becomes
+nonStrictifyB = \case
     TermBind x _ d rhs -> TermBind x NonStrict d rhs
     b                  -> b
 
-mkStrictifyingBinding
-    :: (ToBuiltinMeaning uni fun, PLC.HasUnique name PLC.TermUnique)
-    => BuiltinsInfo uni fun
-    -> VarsInfo tyname name
-    -> Binding tyname name uni fun a
-    -> Maybe (Binding tyname name uni fun a)
-mkStrictifyingBinding binfo vinfo = \case
-    -- Only need to strictify if the previous binding was not definitely pure
-    -- Also, we're reusing the same variable here, see Note [Thunking recursions]
-    TermBind x _ d rhs | not (isPure binfo vinfo rhs) -> Just $ TermBind x Strict d (mkVar x d)
-    _                                                 -> Nothing
+-- Out of a binding `(vardecl x = rhs)`, make a "strictifier" binding: `(strict vardecl x = x)`
+mkStrictifierB :: Binding tyname name uni fun a -> Binding tyname name uni fun a
+mkStrictifierB = \case
+    TermBind x _ d _ -> TermBind x Strict d (mkVar x d)
+    b                -> b
 
 thunkRecursionsStep
-    :: (ToBuiltinMeaning uni fun, PLC.HasUnique name PLC.TermUnique)
+    :: forall tyname name uni fun a b
+    . (ToBuiltinMeaning uni fun, PLC.HasUnique name PLC.TermUnique)
     => BuiltinsInfo uni fun
     -> VarsInfo tyname name
+    -> CompilationOpts b
     -> Term tyname name uni fun a
     -> Term tyname name uni fun a
-thunkRecursionsStep binfo vinfo (Let a Rec bs t) =
+thunkRecursionsStep binfo vinfo opts = \case
+ -- only apply the transformation if there is at least 1 problematic binding in a letrec group
+  Let a Rec bs t | any isProblematic bs ->
     -- See Note [Thunking recursions]
-    let (toThunk, noThunk) = partition strictNonFunctionBinding bs
-        newBindings = fmap nonStrictify toThunk ++ noThunk
-        strictifiers = mapMaybe (mkStrictifyingBinding binfo vinfo) toThunk
-    in mkLet a Rec newBindings $ mkLet a NonRec strictifiers t
-thunkRecursionsStep _ _ t = t
+    let (toNonStrictify, rest) = NE.partition needsNonStrictify bs
+        -- MAYBE: use some prism/traversal to keep the original arrangement of the let group
+        -- this is not a semantic problem, but just a stylistic/debugging issue where the original
+        -- let-group will have reordered the (lazified) bindings
+        editedLet = mkLet a Rec $ fmap nonStrictifyB toNonStrictify ++ rest
+        strictifiers = mkStrictifierB <$> filter needsStrictifier toNonStrictify
+        extraLet = mkLet a NonRec strictifiers
+    in editedLet $ extraLet t
+  t -> t
+  where
+    isStrictEffectful :: Binding tyname name uni fun a -> Bool
+    isStrictEffectful = \case
+        TermBind _ Strict _ rhs -> not $ isPure binfo vinfo rhs
+        _                       -> False
+
+    needsNonStrictify :: Binding tyname name uni fun a -> Bool
+    needsNonStrictify b =
+        if opts^.coPreserveEffectOrder
+        then isProblematic b || isStrictEffectful b
+        else isProblematic b
```

Comment:
I'd like to see some discussion about why it's worth having two modes! Why not just do the conservative thing always? It doesn't look like a lot of savings: the advantage of non-conservative mode is that we keep the recursive binding strict, and so we avoid some overhead from that. But it's not clear to me how much difference that makes. And it only applies in a niche case where we have something that _is_ of function type (so doesn't get strictified anyway), but is impure.

So I'd be pretty inclined to just make the conservative mode the only mode.